### PR TITLE
Add KiKit and dependencies

### PR DESCRIPTION
**Pull request by hesiod**
###### Description of changes

This PR adds [KiKit](https://github.com/yaqwsx/KiKit) as well as its dependencies (all Python packages):
* `pcbnew-transition`
* `pybars3`
* `pymeta3`
* `solidpython`
* `euclid3`

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by hesiod** [comment]
Result of `nixpkgs-review pr 211829` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>python310Packages.euclid3</li>
    <li>python310Packages.pcbnew-transition</li>
    <li>python310Packages.pybars3</li>
    <li>python310Packages.pymeta3</li>
    <li>python310Packages.solidpython</li>
    <li>python311Packages.euclid3</li>
    <li>python311Packages.pcbnew-transition</li>
    <li>python311Packages.pybars3</li>
    <li>python311Packages.pymeta3</li>
    <li>python311Packages.solidpython</li>
  </ul>
</details>

**Comment by hesiod** [comment]
For some reason the addition of `kikit` in `all-packages.nix` didn't get picked up previously (I probably forgot to `git add` it), that's why the rebuild number went up

File: pkgs/development/python-modules/solidpython/default.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  nativeBuildInputs = [ poetry-core pythonRelaxDepsHook ];
+  pythonRelaxDeps = [ "PrettyTable" "pypng" ];
+
+  postPatch = ''
+    substituteInPlace pyproject.toml \
+      --replace "poetry.masonry.api" "poetry.core.masonry.api"
+  '';
+
+  pythonImportsCheck = [ "solid" ];
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/solidpython/default.nix:34]: are we running those tests https://github.com/SolidCode/SolidPython/tree/master/solid/test ?

File: pkgs/development/python-modules/pymeta3/default.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    inherit version;
+    hash = "sha256-GL2jJtmpu/WHv8DuC8loZJZNeLBnKIvPVdTZhoHQW8s=";
+  };
+
+  pythonImportsCheck = [ "pymeta" ];
+
+  checkInputs = [ twisted ];
+
+  # Tests broken on Python 3 (upstream bugs)
+  doCheck = false;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pymeta3/default.nix:22]: FYI python2 is unuspported

File: pkgs/development/python-modules/pybars3/default.nix [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    hash = "sha256-ashH6QXlO5xbk2rxEskQR14nv3Z/efRSjBb5rx7A4lI=";
+  };
+
+  preCheck = ''rm tests.py'';
+  checkInputs = [ unittestCheckHook ];
+
+  propagatedBuildInputs = [ pymeta3 ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pybars3/default.nix:21]:
```suggestion
  propagatedBuildInputs = [ pymeta3 ];

  checkInputs = [ unittestCheckHook ];

  preCheck = ''
    rm tests.py
  '';
```

File: pkgs/development/python-modules/kikit/default.nix [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    click
+    markdown2
+    pybars3
+    lark
+    solidpython
+    commentjson
+    wcwidth
+  ];
+
+  pythonImportsCheck = [ pname ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/kikit/default.nix:51]:
```suggestion
  pythonImportsCheck = [ "kikit" ];
```

File: pkgs/development/python-modules/kikit/default.nix [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+, pybars3
+, pcbnew-transition
+}:
+
+buildPythonPackage rec {
+  pname = "kikit";
+  version = "1.2.0";
+  format = "setuptools";
+
+  # Sources via fetchPypi don't contain tests
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/kikit/default.nix:24]:
```suggestion
```
pretty normal and not worth a comment

File: pkgs/development/python-modules/euclid3/default.nix [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  pname = "euclid3";
+  version = "0.01";
+  format = "setuptools";
+
+  src = fetchPypi {
+    inherit pname version;
+    hash = "sha256-JbgnpXrb/Zo/qGJeQ6vD6Qf2HeYiND5+U4SC75tG/Qs=";
+  };
+
+  pythonImportsCheck = [ pname ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/euclid3/default.nix:15]:
```suggestion
  pythonImportsCheck = [ "euclid3" ];
```

File: pkgs/development/python-modules/euclid3/default.nix [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+  meta = with lib; {
+    description = "2D and 3D vector, matrix, quaternion and geometry module";
+    homepage = "https://pypi.org/project/euclid3/";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/euclid3/default.nix:19]: the real linked website is dead.

**Comment by SuperSandro2000** [comment]
@ofborg build python310Packages.pybars3 python310Packages.kikit python310Packages.pymeta3 python310Packages.kikit python310Packages.euclid3 python310Packages.pcbnew-transition

File: pkgs/development/python-modules/euclid3/default.nix [hunk 8]
```diff
@@ -0,0 +1,24 @@
+{ lib
+, buildPythonPackage
+, fetchPypi
+}:
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/euclid3/default.nix:4]:
```suggestion
}:

```

File: pkgs/development/python-modules/euclid3/default.nix [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  pythonImportsCheck = [ pname ];
+
+  meta = with lib; {
+    description = "2D and 3D vector, matrix, quaternion and geometry module";
+    homepage = "https://pypi.org/project/euclid3/";
+    license = licenses.lgpl21Plus;
+    maintainers = with maintainers; [ tmarkus ];
+  };
+}
+
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/euclid3/default.nix:24]:
```suggestion
```

File: pkgs/development/python-modules/pcbnew-transition/default.nix [hunk 10]
```diff
@@ -0,0 +1,33 @@
+{ lib
+, buildPythonPackage
+, fetchPypi
+, kicad
+, versioneer
+}:
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/pcbnew-transition/default.nix:6]:
```suggestion
}:

```

File: pkgs/development/python-modules/pcbnew-transition/default.nix [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+  ];
+
+  meta = with lib; {
+    description = "Library that allows you to support both KiCAD 5 and KiCAD 6 in your plugins";
+    homepage = "https://github.com/yaqwsx/pcbnewTransition";
+    license = licenses.mit;
+    maintainers = with maintainers; [ tmarkus ];
+  };
+}
+
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/pcbnew-transition/default.nix:33]:
```suggestion
```

File: pkgs/development/python-modules/pybars3/default.nix [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+  ];
+
+  meta = with lib; {
+    description = "Handlebars.js for Python 3 and 2";
+    homepage = "https://github.com/wbond/pybars3";
+    license = licenses.lgpl3Only;
+    maintainers = with maintainers; [ tmarkus ];
+  };
+}
+
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/pybars3/default.nix:34]:
```suggestion
```

File: pkgs/development/python-modules/pymeta3/default.nix [hunk 13]
```diff
@@ -0,0 +1,31 @@
+{ lib
+, buildPythonPackage
+, fetchPypi
+, twisted
+}:
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/pymeta3/default.nix:5]:
```suggestion
}:

```

File: pkgs/development/python-modules/pymeta3/default.nix [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+  doCheck = false;
+
+  meta = with lib; {
+    description = "A Pattern-Matching Language Based on Python";
+    homepage = "https://github.com/wbond/pymeta3";
+    license = licenses.mit;
+    maintainers = with maintainers; [ tmarkus ];
+  };
+}
+
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/pymeta3/default.nix:31]:
```suggestion
```

File: pkgs/development/python-modules/solidpython/default.nix [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+  meta = with lib; {
+    description = "Python interface to the OpenSCAD declarative geometry language";
+    homepage = "https://github.com/SolidCode/SolidPython";
+    license = licenses.lgpl21Only;
+    maintainers = with maintainers; [ tmarkus ];
+  };
+}
+
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/solidpython/default.nix:42]:
```suggestion
```

**Review by AndersonTorres — DISMISSED** [review]

**Comment by hesiod** [comment]
Rebased on top of recent master along with the following changes:
* Update KiKit to 1.3.0 (previously 1.2.0)
* Update pcbnew-transition to 0.3.4 (previously 0.2.0)
* Add solidpython tests
* Add KiKit to `top-level/all-packages.nix` via `toPythonApplication` as it is a hybrid library/application
* Addressed whitespace/style issues from review comments

File: pkgs/development/python-modules/euclid3/default.nix (hunk 7 above)
Comment (hesiod) [reply pkgs/development/python-modules/euclid3/default.nix:20]: Do you mean I should add a comment about this? (the PyPI link works fine)

File: pkgs/development/python-modules/solidpython/default.nix (hunk 1 above)
Comment (hesiod) [reply pkgs/development/python-modules/solidpython/default.nix:41]: Good catch, I added `unittestCheckHook` (and a minor patch) in order to run those tests

**Comment by hesiod** [comment]
@ofborg build python310Packages.kikit python311Packages.kikit

**Comment by hesiod** [comment]
@SuperSandro2000 If you have the time, perhaps you could clarify the two remaining review comments? Otherwise, I think this PR is pretty much good to go :)

File: pkgs/development/python-modules/pymeta3/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/pymeta3/default.nix:23]: We should think about removing the package if it does not support python3

File: pkgs/development/python-modules/euclid3/default.nix (hunk 7 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/euclid3/default.nix:20]: Python packages which have no homepage or source are not that great

**Comment by r0hit05** [comment]
We did not realise that a PR for kikit already exists. We ended up repeating a lot of this work over in #249464. Compared to this PR, it seems like we have added newer versions of certain dependencies. We have also added a kicad package set that can be used with a new `withPackages` helper on kicad.

Sorry for the redundant work!

**Comment by jfly** [comment]
As https://github.com/NixOS/nixpkgs/pull/249464 has landed, I'm going to close this PR.

Again, sorry for the redundant work!

**Pull request closed without merging**